### PR TITLE
[SPARK-40492][SS] Do maintenance before streaming StateStore unload

### PR DESCRIPTION
**Pull request by chaoqin-li1123**
### What changes were proposed in this pull request?
Before unload of a StateStore, perform a cleanup.

### Why are the changes needed?
Current the maintenance of StateStore is performed by a periodic task in the management thread. If a streaming query become inactive before the next maintenance task fire, its StateStore will be unloaded before cleanup. 
There are 2 cases when a StateStore is unloaded.
1. StateStoreProvider is not longer active in the system, for example, when a query ends or the spark context terminates.
2. There is other active StateStoreProvider in the system, for example, when a partition is reassigned.

In case 1, we should do one last maintenance before unloading the instance,

### Does this PR introduce _any_ user-facing change?
Shutdown delay of a query may increase because a maintenance task is scheduled.

### How was this patch tested?
Add an integration test that verify that redundant delta file is deleted when StateStore instances is deactivated and unloaded.

### CODE REVIEW

**Comment by chaoqin-li1123** [comment]
@HeartSaVioR

**Review by HyukjinKwon — COMMENTED** [review]
Let's file a JIRA and link it to the PR title, see also https://spark.apache.org/contributing.html

**Comment by HeartSaVioR** [comment]
I gave a feedback offline but also duplicate here for the history.

`getStateStoreProvider` also report active provider instance and get “inactive provider instances” which could be unloaded without maintenance. 

We can defer these instances to be unloaded till next maintenance interval, but this may not be something we want, since we are effectively reverting SPARK-33827.

Technically, what we really want to do is just to ensure all state store provider instances run maintenance "at least once". That said, ideally we should not perform maintenance on unloading of provider if it has run the maintenance at least once.

In `getStateStoreProvider`, we should only defer some of inactive provider instances which didn't run any maintenance yet. In maintenance task, we need to let provider be unloaded without maintenance if it ran maintenance at least once.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StateStoreSuite.scala [hunk 1]
```diff
@@ -357,6 +357,75 @@ class StateStoreSuite extends StateStoreSuiteBase[HDFSBackedStateStoreProvider]
...
+          generateStoreVersions()
+
+          // If driver decides to deactivate all stores related to a query run,
+          // then this instance should be unloaded.
+          coordinatorRef.deactivateInstances(storeProviderId1.queryRunId)
+          eventually(timeout(timeoutDuration)) {
+            assert(!StateStore.isLoaded(storeProviderId1))
+          }
+
+          // Earliest delta file should be scheduled a cleanup during unload.
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StateStoreSuite.scala:418]: Looks like lots of duplicated code here with "maintenance" and this seems to be the only difference. Do you think we can add the additional check in "maintenance" test instead?

**Review by HeartSaVioR — COMMENTED** [review]
The change looks OK. Only one comment for deduplication.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StateStoreSuite.scala (hunk 1 above)
Comment (chaoqin-li1123) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StateStoreSuite.scala:418]: I thought about it, but "maintenance" test configure maintenance interval to be very small, which make it hard to verify the behavior.(we want maintenance interval to be larger than query lifetime in this case)

Comment (HeartSaVioR) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StateStoreSuite.scala:418]: I see. If we are relying on timing of maintenance interval, please explicitly set the interval to ensure that the test won't fail on changing default. Also please leave code comment about the condition of maintenance interval to make this test be successful.

Btw, would you mind if I ask to run this test case 100 times and verify we don't have flakiness?

Thanks in advance!

Comment (chaoqin-li1123) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StateStoreSuite.scala:418]: Sure, I will run the test 100 times. This test is deterministic from my understanding. After we verify that unload is called, maintenance must also have been called. The reason why a shorter interval is not good for this test is that we can't differentiate last maintenance of unloaded instances and normal maintenance of active instances.

**Review by HeartSaVioR — APPROVED** [review]
+1

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master.

**Comment by HeartSaVioR** [comment]
Thanks @chaoqin-li1123 for the contribution! I merged this to master.

**Pull request closed without merging**